### PR TITLE
[Personal WIP] Story of Layout feature

### PR DESCRIPTION
**Pull request by albertusdev**
A PR to show my progress on working on this feature.

Dependency:
- Flutter framework change:
https://github.com/albertusangga/flutter/pull/2

### CODE REVIEW

**Comment by albertusdev** [comment]
Check-in 1:
- Show BoxConstraints next to the summary tree
- Highlight constraints of widget that unconstrained it's child

image

**Comment by albertusdev** [comment]
Check-in 2:
- Rename Show Constraints button to Layout button (change term of showConstraints to debugLayoutModeEnabled)
- Change color from red to yellow (Dangerous to Warning, since we are unsure whether it's really an error)
- Highlight children that got unconstrained by its parents (highlight the contraints text)
- Highlight the parent widget name that unconstrained its child
- Show fade-in/fade-out animation when Debug Layout mode 

Concerns:
- performance on Flutter web since I'm using ValueNotifier & ValueListenableBuilder for all of the inspector row nodes

image

**Comment by albertusdev** [comment]
Check-in 3:
- Made very early prototype on explaining Flex layouting (the feature for now will be temporarily called 'Story of Your Flex Widget')
- Visualizing Flex and able to detect the Flex type (Row/Column)
- Displaying correct amount of children and all of the children's widget name
- All children are still equally weighted for this check in

image
image

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 import 'dart:math';
 
+import 'package:devtools_app/src/ui/fake_flutter/_real_flutter.dart';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: auto imports are doing the wrong thing. just go ahead and import 
'package:flutter/material.dart'

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 2]
```diff
@@ -617,22 +618,80 @@ class _ConstraintsDescriptorState extends State<ConstraintsDescriptor>
...
   }) : super(key: key);
 
+  InspectorTreeNode node;
+
   @override
   Widget build(BuildContext context) {
+    final Map<String, Object> flexDetails = node.diagnostic?.json['flex'];
+    final int numOfChildren = node.diagnostic.childrenNow.length;
+    final List<Widget> children = node.diagnostic.childrenNow.map((RemoteDiagnosticsNode child) => Expanded(
+      flex: 1,
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: you can now use List comprehensions for this sort of case in dart. so
[for (var child in node.diagnostic.childrenNow) 
   Container(something something)
];

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 3]
```diff
@@ -617,22 +618,80 @@ class _ConstraintsDescriptorState extends State<ConstraintsDescriptor>
...
-          Expanded(
-            child: Center(
-              child: Text('TODO'),
+      child: Container(
+        margin: const EdgeInsets.all(16.0),
+        child: Column(
+          children: [
+            Container(
+              margin: const EdgeInsets.only(bottom: 4.0),
+              child: const Text('Story of Your Flex Widget', style: TextStyle(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: Add the name of the widget class. So:
'Story of the flex layout of your ${diagnostic.description} widget'
so
Story of the flex layout of your Column widget
or
Story of the flex layout of your Row widget.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 4]
```diff
@@ -617,22 +618,80 @@ class _ConstraintsDescriptorState extends State<ConstraintsDescriptor>
...
-  const StoryOfYourFlexWidget(InspectorTreeNode node, {
+  StoryOfYourFlexWidget(this.node, {
     Key key,
   }) : super(key: key);
 
+  InspectorTreeNode node;
+
   @override
   Widget build(BuildContext context) {
+    final Map<String, Object> flexDetails = node.diagnostic?.json['flex'];
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: Exciting to see this!
Go ahead and add getters for the extra json fields you add to RemoteDiagnosticsNode.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 5]
```diff
@@ -617,22 +618,80 @@ class _ConstraintsDescriptorState extends State<ConstraintsDescriptor>
...
-        ],
+            Center(
+              child: Text('${node.diagnostic?.json['flex']}'),
+            ),
+            Center(
+              child: Text('${node.diagnostic?.childrenNow}'),
+            ),
+            Expanded(
+              child: Container(
+                color: Colors.blue,
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: try to use colors from the Theme whenever possible. That will make the ui work in dark and light theme.
See  calls to Theme.of in the code for examples.

**Comment by albertusdev** [comment]
Check in 4:
- Avoid using hardcoded color and instead use Color from Theme (to support Light/Dark theme)
- Remove hack in InspectorTree node
- Remove wrong import
- Rename 'Story of your Flex widget' to 'Story of the layout of your Row/Column widget'

TODOs for tomorrow:
- Choose more contrasts and sutiable colors
- Display Flex widget size, constraints, main axis and cross axis
- Properly size Flex children

Screenshots:
image

**Comment by albertusdev** [comment]
## Check in 5:

### Changelist:
- Create proper deserializer method for flex properties from DiagnosticsNode
- Create new flag (`isExperimentalStoryOfLayoutEnabled`) in `InspectorTreeFlutterController` to prevent showing all this features changes until my PR (github.com/albertusangga/flutter/pull/2) is landed to Flutter main repository
- Inspector details tree now live inside of Tabs! User can focus on layout details instead if they want to.
- Removed animation from DianogsticsNodeDescriptions for now (due to minor performance concerns and adding new flag, might add it later!)
- Changed hardcoded constants for Buttons in InspectorScreen to prevent button groups expanding to two lines 

### Screenshots:
- Inspector screen have tabs now!
image

### TODOs:
- Remove `// constraints` word from ConstraintsDescriptor
- Highlight only unconstrained word in the ConstraintsDescriptor
- Move StoryOfYourFlexWidget from Dialog to Tab

**Pull request closed without merging**